### ISSUE TITLE
Syntax Suggestions for C# 8 Await Foreach and Extensions

### ISSUE THREAD

**Issue by siegfriedpammer** [dotnet/csharplang#840]
I just watched the video on Channel 9 about the new C# 8 features. I really like the new features!
I hope I am not too late, but there are just two things I would like to propose:

1. `foreach await (var s in stream)` looks a bit strange. How about `foreach (await var s in stream)`? This would make it more C#-like as we have only one keyword before the open parenthesis in all the other control structures.
2. `extension A extends ClassA : ...` is very similar to inheritance in other languages (Java), which might be confusing for other people. Plus, it's a new keyword. I would propose to use `extension A for ClassA : ...`.
Benefits include:
a) Reuse of existing keyword.
b) Similar to natural language.

**Comment by DavidArno** [comment]
@siegfriedpammer,

Could you share a link to the video, please? It helps those of us who haven't seem it and puts context to your suggestions.

**Comment by benaadams** [comment]
https://channel9.msdn.com/Blogs/Seth-Juarez/A-Preview-of-C-8-with-Mads-Torgersen

**Comment by flq** [comment]
1. Agree - Either that, or possibly `foreach (await var s in stream)`
2. Just pushing it out there, brainstorm style, what about `extension A<Customer> { }` ? Constrains the type of this to be the "generic argument"

**Comment by FurkanKambay** [comment]
Maybe `await foreach(string s in myStream)`? I think it helps with readability; you know, like "await for each string s in myStream).

**Comment by mariusGundersen** [comment]
`foreach await(string s in myStream)` is similar to [EcmaScript stage 3 proposal](https://github.com/tc39/proposal-async-iteration) of `for await(let s of myStream)`, so I would recommend keeping that syntax.

**Comment by siegfriedpammer** [comment]
@mariusGundersen Is it me, or is JavaScript creeping into C#?
@flq `extension A<Customer> ...` would work for me as well, I think using a keyword would make it more obvious, what's going on, though.

**Comment by mariusGundersen** [comment]
@siegfriedpammer, as Mads said in the video, they are stealing ideas from TypeScript, which is a superset of EcmaScript, so I guess ideas are creeping in both directions.

**Comment by siegfriedpammer** [comment]
@mariusGundersen I am fine with ideas and features, but I don't think C# should follow the example of JavaScript in terms of syntax.

**Comment by eyalsk** [comment]
@siegfriedpammer 

> extension A extends ClassA : ... is very similar to inheritance in other languages (Java), which might be confusing for other people. Plus, it's a new keyword. I would propose to use extension A for ClassA : ....
> Benefits include:
> a) Reuse of existing keyword.
> b) Similar to natural language.

Just because the semantics for `extends` are different in other languages doesn't mean that C# shouldn't use it as a keyword and instead look for another confusing word such as `for` when it may imply that you extend a set of classes as opposed to a single one.

@mariusGundersen 

> foreach await(string s in myStream) is similar to EcmaScript stage 3 proposal of for await(let s of myStream), so I would recommend keeping that syntax.

I really don't like it, I don't think you `await` the whole thing and prefer to have `foreach (string s await myStream)`.

**Comment by MillKaDe** [comment]
@eyalsk 
`await` should not placed right before the `IEnumerable<T> myStream`, because it is not the stream as a whole that is awaited. Instead each item pulled out (by some MoveNext-like method) is awaited. There is one `await` for each item, not one `await` for the whole stream of items.

For the same reason, `await foreach (..)` doesn't make sense.  It is not the whole loop, that is awaited.

IMHO, the syntax should be one of
1. `foreach await (var i in stream) ...` (as proposed by @MadsTorgersen )
2. `foreach (await var i in stream) ...` (as proposed by @siegfriedpammer above)
3. `foreach (var await i in stream) ...`

**Comment by benaadams** [comment]
`foreach (var i await in stream) ...`

>  It is not the whole loop, that is awaited.

Is an async foreach `Task` returning?

e.g.
````
Task t = foreach await (var i in stream) ....

// ...

await t;
```

**Comment by pentp** [comment]
`foreach (await var s in stream)` feels better that the alternatives because `await` is "executed" for each variable pulled from `stream`, not on the `foreach` as a whole. It doesn't make much sense between the variable type and name declarations and can't be after `in` because that's valid syntax today.

**Comment by benaadams** [comment]
Adding an cosmetic equals for variable assignment

`foreach (var i = in stream) ...` (current `foreach`) 

1. `foreach await (var i = in stream) ... `
2. `foreach (await var i = in stream) ... `
3. `foreach (var await i  = in stream) ...`
4. `foreach (var i = await in stream) ....`

I still think 4. `foreach (var i await in stream)` makes most sense as `i` is the awaited result not the task that you then await in the loop.

**Comment by MillKaDe** [comment]
@benaadams 
Your option 4 makes my eyes bleed a little bit ... ;-P

But it is maybe the best syntax variant to avoid any misunderstandings about whats going on.  The `await` happens when/while pulling the next item out of the stream, and when the `await` completes, the pulled item is assigned to `i`.

**Comment by benaadams** [comment]
> Your option 4 makes my eyes bleed a little bit ... 

@MillKaDe I assume there is an implicit await on the foreach otherwise the options are kinda

1. `await foreach await (var i = in stream) ... `
2. `await foreach (await var i = in stream) ... `
3. `await foreach (var await i  = in stream) ...`
4. `await foreach (var i = await in stream) ...`

As I assume you should be able to capture the overall sequence in a Task

1. `Task t = foreach await (var i = in stream) ... `
2. `Task t = foreach (await var i = in stream) ... `
3. `Task t = foreach (var await i  = in stream) ...`
4. `Task t = foreach (var i = await in stream) ...`

**Comment by siegfriedpammer** [comment]
@benaadams I am not sure whether the assignment in the middle is necessary or adds anything. The current foreach is just `foreach (var item in collection) ...`.

**Comment by benaadams** [comment]
@siegfriedpammer yeah was just for demonstration of what was happening with the variable assignment; removed from follow up comment, about the foreach itself having an implicit await; and if its Task-like can be captured.

**Comment by MillKaDe** [comment]
@benaadams 
>I assume there is an implicit await on the foreach ..

I don't think so - As far as I understand the planned feature, the foreach loop as a whole is not awaitable.  
Of course, the foreach loop (and optional stuff before and after) can be put in `Task` / async function, which can be awaited.

**Comment by benaadams** [comment]
> As far as I understand the planned feature, the foreach loop as a whole is not awaitable.

That means it can only be used in an `async` function? So you couldn't do

```csharp
    Task DoThing(IAsyncEnumerable e)
    {
        return foreach (var i await in e)
        {
            //
        }
    }
```
But would have to do:
```csharp
    async Task DoThing(IAsyncEnumerable e)
    {
        foreach (var i await in e)
        {
            //
        }
    }
```

**Comment by MillKaDe** [comment]
@siegfriedpammer 
>I am not sure whether the assignment in the middle is necessary or adds anything.

Yep, thats what made my eyes bleed, because a classical foreach doesn't has a `=`.
But the classical `await` has it: `var x = await y;`
The explicit `=` in Ben's suggested syntax makes it crystal clear, that awaiting happens on the right side of `=` when the next item is pulled out of the stream, and that after that operation has completed, its result is assigned to `i`.

Is is necessary to be so explicit ? I don't know. I messed around with async/await a while ago and had to analyze how it works exactly. So I don't need that explicit hint about where the `await` happens.

But those, who just use async/await, might get wrong ideas without that hint.

**Comment by eyalsk** [comment]
@MillKaDe Okay, I see what you mean and I agree. 😄

**Comment by stephentoub** [comment]
> That means it can only be used in an async function?

Yes, just as await can only be.  What are you concerned about, @benaadams?

**Comment by benaadams** [comment]
> What are you concerned about,

Thinking it though the awaits would be unwrapped directly into the function; so you would need a Task to represent the sum of them, so being in an async function that represented them would be fine.

I was originally thinking that there was already a sum Task that represented the foreach so adding a function for only a foreach would carry a double sum Task overhead. However, I see that's not the case now.

**Comment by jnm2** [comment]
I like `foreach await`.

I was hoping for Shapes before Extension Everything, though. If they weren't mentioned, does that mean they'll likely be a 9.0 feature?

**Comment by eyalsk** [comment]
@jnm2 I'd be really surprised if we gonna have shapes for 8.0 or 8.x and yeah after what @MillKaDe said, `foreach await` make a lot more sense to me. :)

**Comment by gulshan** [comment]
I don't know if naming extension classes is necessary. If we avoid naming, the syntax may look just like-
```csharp
extension String
{
    ...
}
```
Swift is using this syntax.

**Comment by CyrusNajmabadi** [comment]
> I don't know if naming extension classes is necessary

In general we like to name things that are outside a method-body's scope.  This is because we have to consider things like interop of our libraries.  Both with previous versions of the language (i.e. someone using C#7 and earlier), and other languages (i.e. F#/VB/etc.).  Language that choose to not incorporate special knowledge of these types still need ways for their users to get to that functionality.  

That's why, for example, extension methods today are inside static classes.  That enables anyone to get to them like normal methods.

**Comment by gulshan** [comment]
@CyrusNajmabadi Can the naming be optional and a name will be auto-generated by convention when no name given?  Something like-
```csharp
//Name auto-generated to StringExtension or something like that
extension String
{
    ...
}

//Name not generated
static class MyStringExtension extension String
{
    ...
}
```

The if any conflict happens with convention based auto generated names, an error will be shown.

**Comment by Joe4evr** [comment]
If you let the compiler generate a name, then it's always an "unspeakable" name in order to avoid conflicts. I'd rather not change that behavior, since otherwise that would need an exact specification on how to generate a name and you're marrying yourself to that spec for the rest of time.

**Comment by artelk** [comment]
I've seen that the proposed design of the interfaces is like this:

```c#
public interface IAsyncEnumerable<T>
{
  public IAsyncEnumerator<T> GetEnumerator();
}

public interface IAsyncEnumerator<T>
{
  public T Current { get; }
  public Task<bool> MoveNextAsync();
}
```

I believe it is MUCH better to change the IAsuncEnumerator to this:
```c#
public interface IAsyncEnumerator<T>: IDisposable
{
  public Task<Optional<T>> MoveNextAsync();
}
```
And assume the the None value is returned at the end.
This would allow to parallelize the things.

```c#
Task<Something> MyOperationAsync(obj)
{
   var x = await DoSomethingSlowAsync(obj);
   await WriteToDBAsync(x);
   return new Something(obj, x);
}

Task<...> MySecondOperationAsync(...)
{
  ///...
}

var results = GetStreamOfObjectFromRemoteService()
   .Select(o => MyOperationAsync(o))
   .Select(o => MySecondOperationAsync(o))
   .Parallelize(maxDegreeOfParallelizm: 20, maxPrefetchedItemsQueueSize: 1000) // this parallelizes everything above
   .ToListAsync();
```
Assuming that not only the MySecondOperationAsync is parallelized. I mean this should work the same as
```c#
var results = GetStreamOfObjectFromRemoteService()
   .Select(o =>
               {
                   var a = await MyOperationAsync(o);
                    return await MySecondOperationAsync(a);
               })
   .Parallelize(maxDegreeOfParallelizm: 20, maxPrefetchedItemsQueueSize: 1000)
   .ToListAsync();
```

So each time you call MoveNextAsync() you potentially start a new task.
This could be very useful in practice to have such parallelization feature.
If the Current is a separate mutable property you wan't be able to use it in parallel.

Of course the enumerators generated by using async yield returns should put semaphore.WaitAsync() to prevent from parallelization. But the methods like Select should allow executing the continuations in parallel if the client called MoveNextAsync multiple times not awaiting the previous calls results.


Btw, I think the Linq methods can be implemented directly for the IAsyncEnumerator.
Maybe IAsyncEnumerable is not needed at all.

**Comment by HaloFour** [comment]
@artelk 

IIRC the problem with defining the interface like that is the loss of covariance.

**Comment by jnm2** [comment]
While we're at it, how about this?

```c#
public interface IAsyncEnumerator<T>: IDisposable
{
  public Task<IReadOnlyList<T>> MoveNextAsync();
}
```

Allows for batching. Empty list signals end of async enumeration.

**Comment by artelk** [comment]
@HaloFour 
That's true. But maybe we would also like to use ValueTask instead of Task? So anyways we lose covariance.
We could implement an Upcast<TBase> operation (safe version of Cast allowing only upcasts).

**Comment by artelk** [comment]
@jnm2 
Is it supposed to preload all items before returning the list or return only currently fetched/available ones?
What is the marker of EndOfStream? Return null?

**Comment by wholroyd** [comment]
For the conversation around Extensions syntax, I don't know why we wouldn't try to limit the introduction of keywords. The addition of `extension` is understood, but not the addition of `extends`. If what I understood by the Channel9 video - extensions are classes on their own right (since they would be able to maintain their own state via statics), then the syntax and behavior of the extension class should try to follow the syntax of it just the same in my opinion.

Current syntax for extensions on a reference type...

```csharp
public static class PersonExtensionClass
{
  public string FullName(this Person person)
  {
    return $"{person.FirstName} {person.MiddleName} {person.LastName}";
  }
}
```

Current proposed syntax for extensions on a reference type...

```csharp
public extension PersonExtensionClass extends Person
{
  public string FullName()
  {
    return $"{this.FirstName} {this.MiddleName} {this.LastName}";
  }
}
```

My proposed syntax for extensions on a reference type is mainly around the class declaration itself, not the logic being suggested within the class definition itself. My main gripe is that the proposal is for extensions to become their own classes in itself, as they would also maintain their own state, aren't following the syntax of normal classes. If extensions should be able to extend other extensions, there is the requirement to inheir like normal classes.

```csharp
public extension PersonExtensionClass<IPerson>
{
  // code
}
public extension PersonExtensionClassOfExtensionClass<IPerson> : PersonExtensionClass<IPerson>
{
  // code
}

public extension PersonExtensionClassOfExtensionClass<IPerson> : PersonExtensionClass<IPerson>, IDispose
{
  // code
}
```

Seems a bit more natural in this usage to me.

**Comment by jnm2** [comment]
> Is it supposed to preload all items before returning the list or return only currently fetched/available ones?

The latter. It waits for at least one item and returns all the items it has at that time without waiting for more.

> What is the marker of EndOfStream? Return null?

Like I said, empty list signals end of async enumeration. I think `null` is not a good idea.

**Comment by jnm2** [comment]
:D https://twitter.com/onovotny/status/900482078537654274

**Comment by artelk** [comment]
@HaloFour
Imagine our Extension feature is so powerful so you can specify an implicit cast from IAsyncEnumerator<TDerived> to  IAsyncEnumerator<TBase> by wrapping it into an adapter.
Or you could even avoid the adapter and provide an implementation of IAsyncEnumerator[TBase] extending the IAsyncEnumerator[TDerived]. I mean something like this:
```c#
public extension IAsyncEnumeratorExtensions<TDerived, TBase> where TDerived: TBase for IAsyncEnumerator<TDerived>: IAsyncEnumerator<TBase>
{
    public Task<Optional<TBase>> MoveNextAsync()
    {
        Optional<TDerived> item = await this.MoveNextAsync();
        return item.IsNone ? Optional<TBase>.None : new Optional<TBase>(item.Value);
    }
}
```
So even Object.ReferenceEquals would return true.
This requires some changes in CLR. But they were mentioned in the presentation. Not sure if that supposed to be so flexible to support such generic extensions.

**Comment by IEvangelist** [comment]
I wanted to add to the conversation regarding the "extension everything" syntax. Current syntax adds two new keywords:

 - `extension`
 - `extends`

I think we really only need one. The `extension` keyword is really just an alias for `static class` that is intended to be an extension class. The real work horse it would seem is the `extends` keyword. I'd like to propose having this instead:

```csharp
public class Foo extends Bar
{
   // ... extension everything stuff
}
```
It's really a tiny change, but avoids two new keywords when one will do. Now, wherever you see `extends` this is an extension class.

**Comment by artelk** [comment]
I would suggest the following design for IAsyncEnumerator:

```c#
public interface IAsyncEnumerator<out T> : IDisposable // covariant
{
    ValueTask<ResultToken?> MoveNextAsync();
    T FetchResult(ResultToken resultToken);
}

public struct ResultToken
{
    internal readonly object source;
    internal readonly long key;
    internal ResultToken(object source, long key)
    {
        this.source = source;
        this.key = key;
    }
}

public static class AsyncEnumerator
{
    public static async ValueTask<Optional<T>> FetchNextAsync<T>(this IAsyncEnumerator<T> source)
    {
        ResultToken? token = await source.MoveNextAsync().ConfigureAwait(false);
        return token.HasValue ? source.FetchResult(token.Value) : Optional<T>.None;
    }
}

public struct Optional<T>
{
    public static readonly Optional<T> None = default(Optional<T>);

    private readonly bool hasValue;
    private readonly T value;

    public Optional(T value)
    {
        this.value = value;
        this.hasValue = true;
    }

    public bool HasValue => this.hasValue;

    public T Value => this.hasValue ? this.value : throw new InvalidOperationException("Doesn't have value");

    public static implicit operator Optional<T>(T value) => new Optional<T>(value);
}

internal abstract class AsyncEnumeratorBase<T> : IAsyncEnumerator<T>
{
    private readonly ConcurrentDictionary<long, T> results = new ConcurrentDictionary<long, T>();
    private long lastKey;

    public async ValueTask<ResultToken?> MoveNextAsync()
    {
        Optional<T> result = await this.FetchNextAsync().ConfigureAwait(false); // virtual method
        if (!result.HasValue)
        {
            return null;
        }

        long key = Interlocked.Increment(ref this.lastKey);
        this.results[key] = result.Value;
        return new ResultToken(this, key);
    }

    public T FetchResult(ResultToken resultToken)
    {
        if (!ReferenceEquals(resultToken.source, this))
        {
            throw new ArgumentException("Token from different source");
        }

        return this.results.TryRemove(resultToken.key, out T result)
            ? result
            : throw new InvalidOperationException("Has been already fetched");
    }

    protected abstract ValueTask<Optional<T>> FetchNextAsync();

    public abstract void Dispose();
}

//Sample async enumerator
internal sealed class WhereAsyncEnumerator<T> : AsyncEnumeratorBase<T>
{
    private readonly IAsyncEnumerator<T> source;
    private readonly Func<T, bool> predicate;

    public WhereAsyncEnumerator(IAsyncEnumerator<T> source, Func<T, bool> predicate)
    {
        this.source = source;
        this.predicate = predicate;
    }

    protected override async ValueTask<Optional<T>> FetchNextAsync()
    {
        while (true)
        {
            Optional<T> result = await this.source.FetchNextAsync().ConfigureAwait(false); // extension method
            if (!result.HasValue || this.predicate(result.Value))
            {
                return result;
            }
        }
    }

    public override void Dispose()
    {
        source.Dispose();
    }
}
```

**Comment by tyrotoxin** [comment]
@artelk that design looks impractical. You can look at [AsyncEnumerable library](https://github.com/tyrotoxin/AsyncEnumerable) - the implementation can be more complex than that.

**Comment by btecu** [comment]
I'm guessing this can be closed as C# shipped with `await foreach (...)`.
